reactive/etcd.py

```diff
@@ -97,13 +105,53 @@ def update_cluster_string():
 
 @when_not('etcd.installed')
 def install_etcd():
-    source = config('source-url')
-    sha = config('source-sum')
-
     status_set('maintenance', 'Installing etcd.')
-    etcd_helper = EtcdHelper()
-    etcd_helper.fetch_and_install(source, sha)
-    set_state('etcd.installed')
+
+    codename = host.lsb_release()['DISTRIB_CODENAME']
+
+    etcd_path = resource_get('etcd')
+    etcdctl_path = resource_get('etcdctl')
+
+    if not etcd_path or not etcdctl_path:
+        if codename == 'xenial':
+            # edge case where archive allows us a nice fallback on xenial
+            status_set('maintenance', 'Attempting install of etcd from apt')
+            pkg_list = ['etcd']
+            apt_update()
+            apt_install(pkg_list, fatal=True)
+            set_state('etcd.installed')
+            return
+        else:
+            # edge case
+            status_set('blocked', 'Missing Resource: see README')
+    else:
+        install(etcd_path, '/usr/bin/etcd')
+        install(etcdctl_path, '/usr/bin/etcdctl')
+
+        host.add_group('etcd')
+
+        if not host.user_exists('etcd'):
+            host.adduser('etcd')
+            host.add_user_to_group('etcd', 'etcd')
+
+        os.makedirs('/var/lib/etcd/', exist_ok=True)
```

Comment:
exist_ok is only valid in 3.2 and up, making this a 2.7 incompatible change. I see you changed the shebang to python3 so this is fine, but it takes only one more line to make it compatible with 2.7 (here)
```python
if not os.isdir('/var/lib/etcd')
    os.makedirs('/var/lib/etcd)
```